### PR TITLE
All optimizations

### PR DESCRIPTION
**Pull request by jltien**
Put the optimizations from the Optimizations-1-NoPrefercapes-RMP-EarlyEnd-LateStart with the algorithm and added the other optimizations (least days, most days, least gaps, most gaps).

### CODE REVIEW

File: ucsdscheduleme/Repo/ScheduleRepo.cs [hunk 1]
```diff
@@ -210,5 +210,476 @@ private bool IsWithinTime(List<Section> schedule, int Earliest, int Latest)
...
         }
+
+        /// <summary>
+        /// Returns the schedule with the highest received GPAs from CAPEs.
+        /// </summary>
+        /// <returns>The schedule with highest CAPEs GPA</returns>
+        /// <param name="possibleSchedules">Possible schedules with no time conflicts.</param>
+        public List<Section> HighestGPA(List<List<Section>> possibleSchedules)
+        {
+            List<Section> result = possibleSchedules[0];
```
Comment (git-ram) [inline ucsdscheduleme/Repo/ScheduleRepo.cs]: A little comment to explain this line is helpful

File: ucsdscheduleme/Repo/ScheduleRepo.cs [hunk 2]
```diff
@@ -210,5 +210,476 @@ private bool IsWithinTime(List<Section> schedule, int Earliest, int Latest)
...
+        /// <returns>The schedule with the least days.</returns>
+        /// <param name="possibleSchedules">Possible schedules with no time conflicts.</param>
+        public List<Section> LeastDays(List<List<Section>> possibleSchedules)
+        {
+            int leastDay = 5;
+            List<Section> leastDaySchedule = possibleSchedules[0];
+
+            // Iterate through the possible schedules and counting the number of days
+            // per schedule
+            foreach (List<Section> schedule in possibleSchedules)
```
Comment (git-ram) [inline ucsdscheduleme/Repo/ScheduleRepo.cs]: Make it a separate function:D

File: ucsdscheduleme/Repo/ScheduleRepo.cs [hunk 3]
```diff
@@ -210,5 +210,476 @@ private bool IsWithinTime(List<Section> schedule, int Earliest, int Latest)
...
+            {
+                List<List<Tuple<int, int>>> week = new List<List<Tuple<int, int>>>();
+
+                foreach (Section section in schedule)
+                {
+                    foreach (Meeting meeting in section.Meetings)
+                    {
+                        if (meeting.Days.HasFlag(Days.Monday))
+                        {
+                            Tuple<int, int> time = new Tuple<int, int>(meeting.StartTime, meeting.EndTime);
```
Comment (harishduwadi) [inline ucsdscheduleme/Repo/ScheduleRepo.cs]: We can create tuple only once before all the if statements.

File: ucsdscheduleme/Repo/ScheduleRepo.cs [hunk 4]
```diff
@@ -210,5 +210,476 @@ private bool IsWithinTime(List<Section> schedule, int Earliest, int Latest)
...
+            foreach (List<Section> schedule in possibleSchedules)
+            {
+                List<List<Tuple<int, int>>> week = new List<List<Tuple<int, int>>>();
+
+                foreach (Section section in schedule)
+                {
+                    foreach (Meeting meeting in section.Meetings)
+                    {
+                        if (meeting.Days.HasFlag(Days.Monday))
+                        {
```
Comment (harishduwadi) [inline ucsdscheduleme/Repo/ScheduleRepo.cs]: Meeting was of different types, so was counting for the meeting type of final and reviews as well.

**Pull request merged**